tests/test_timer.cpp

```diff
@@ -37,116 +37,112 @@ struct Timer : public testing::Test
     std::atomic_bool timedout = false;
     std::condition_variable cv;
     std::mutex cv_m;
+    const std::chrono::milliseconds default_timeout{10};
+    const std::chrono::milliseconds test_sleep{20};
+    const std::chrono::seconds default_wait{1};
 };
 } // namespace
 
 TEST_F(Timer, times_out)
 {
-    mpu::Timer t{duration_cast<seconds>(1ms), [this]() {
-                     {
-                         std::lock_guard<std::mutex> lk{cv_m};
-                         timedout.store(true);
-                     }
+    mpu::Timer t{default_timeout, [this]()
+                 {
+                     timedout.store(true);
```

Comment:
Why remove the lock?

Reply:
Having the lock in place here can potentially produce a false positive result in that a timeout could actually occur and call this callback, but `timedout` won't get set until the lock is acquired and the check in the test itself will pass because this is blocked.  I observed this tracing the code.  I hope this makes sense, but it's quite complicated.  We can follow up on chat if it will help.

Reply:
I had another look. According to [this](https://en.cppreference.com/w/cpp/thread/condition_variable), the lock is needed for this to operate properly:

> Even if the shared variable is atomic, it must be modified under the mutex in order to correctly publish the modification to the waiting thread.

Reply:
That said, I think these tests make a few assumptions that may not hold. One issue is line 60 assuming no spurious wakes. Assuming there is no timeout between 54 and 55 is another. Between that and 58 yet another. Locks would be needed for all of those assumptions to work. Here's how I think the test should look:

```cpp
TEST_F(Timer, times_out)
{
    mpu::Timer t{default_timeout, [this]()
                 {
                     {
                         std::unique_lock<std::mutex> lk(cv_m);
                         timedout.store(true);
                     }
                     cv.notify_all();
                 }};

    std::unique_lock<std::mutex> lk(cv_m);
    t.start();
    ASSERT_FALSE(timedout.load()) << "Should not have timed out yet"; // because this is locked

    ASSERT_TRUE(cv.wait_for(lk, default_wait, [this] { return timedout.load(); }))
        << "Did not time out within reasonable timeframe";
}
```

The main thread needs to synchronize with the timer during the whole time it is running, hence the lock before start (and in the timer). That lock is only released by the condition variable itself. I pass a predicate to the wait operation to deal with spurious wakes and check the return for timeouts.

Reply:
Ok, I agree that we need a predicate in the `wait_for` and also agree that we need to lock the mutex in the callback.  What I don't agree with, and found during fixing this, is acquiring the lock before calling `start()` and holding it the whole time before calling `wait_for()`.

The reason being is that if the lock is held this whole time, it's still possible that the timer times out, but it will block in the callback waiting for the mutex before it can set `timedout` to `true`.  However, the main test thread will see that `timedout` is `false` (the `ASSERT_FALSE(timedout.load()) << "Should not have timed out yet";`) which will result in a false positive since it really did timeout, it's just that it couldn't set the `bool` yet.